azurelinuxagent/ga/exthandlers.py

```diff
@@ -329,36 +322,40 @@ def format_value(parse_fn, value):
         return activity_id, correlation_id, gs_creation_time
 
     def run(self):
-
         try:
             self.ext_handlers, etag = self.protocol.get_ext_handlers()
-            msg = u"Handle extensions updates for incarnation {0}".format(etag)
-            logger.verbose(msg)
-            # Log status report success on new config
-            self.log_report = True
-            incarnation_changed = self._incarnation_changed(etag)
-            if self._extension_processing_allowed() and incarnation_changed:
-                activity_id, correlation_id, gs_creation_time = self.get_goal_state_debug_metadata()
-
-                logger.info(
-                    "ProcessGoalState started [Incarnation: {0}; Activity Id: {1}; Correlation Id: {2}; GS Creation Time: {3}]".format(
-                        etag, activity_id, correlation_id, gs_creation_time))
 
-                self.__process_and_handle_extensions(etag)
-                self.last_etag = etag
+            if not (self._extension_processing_allowed()):
+                return
 
-            self.report_ext_handlers_status(incarnation_changed=incarnation_changed)
-            self._cleanup_outdated_handlers()
+            activity_id, correlation_id, gs_creation_time = self.get_goal_state_debug_metadata()
+
+            logger.info(
+                "ProcessExtensionsInGoalState started [Incarnation: {0}; Activity Id: {1}; Correlation Id: {2}; GS Creation Time: {3}]".format(
+                etag, activity_id, correlation_id, gs_creation_time))
+
+            utc_start = datetime.datetime.utcnow()
+            try:
+                self.__process_and_handle_extensions(etag)
+                self._cleanup_outdated_handlers()
+            finally:
+                duration = elapsed_milliseconds(utc_start)
+                msg = 'ProcessExtensionsInGoalState completed [Incarnation: {0}; {1} ms; Activity Id: {2}; Correlation Id: {3}; GS Creation Time: {4}]'.format(
+                    etag, duration, activity_id, correlation_id, gs_creation_time)
+                logger.info(msg)
+                add_event(
+                    AGENT_NAME,
+                    op=WALAEventOperation.ProcessGoalState,
+                    duration=duration,
+                    message=msg)
         except Exception as error:
-            msg = u"Exception processing extension handlers: {0}".format(ustr(error))
-            detailed_msg = '{0} {1}'.format(msg, traceback.extract_tb(get_traceback(error)))
+            msg = u"ProcessExtensionsInGoalState - Exception processing extension handlers: {0}\n{1}".format(ustr(error), traceback.extract_tb(get_traceback(error)))
```

Comment:
Previously the order of Kusto events used to be - 
Started -> Error -> Ended

Now, it would look something like - 
Started -> Ended -> Error

I have a doubt this will confuse the support people a bit. Any way we can keep the original sequence without too much troubles?

Reply:
would this be just moving the above `finally` down to this `try: except`?

Reply:
One problem I could see scope of those variables if we move finally outside of main try:except

Reply:
@larohra yes, i was aware of the change but felt it was not significant; since you think this may be confusing i'll go ahead and change this.

@kevinclark19a @nagworld9 this requires a little more code massaging, we should emit only 1 message. Initially I wanted to avoid that since I felt the change is not significant, but I will go ahead and do it

Reply:
see latest commit for those changes, the code changed quite a bit in order to handle errors correctly

Reply:
Thanks!